File: android-sdk/src/main/java/co/elastic/apm/android/sdk/internal/features/centralconfig/CentralConfigurationManager.java
/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *	http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package co.elastic.apm.android.sdk.internal.features.centralconfig;

import android.content.Context;

import androidx.annotation.VisibleForTesting;

import com.dslplatform.json.DslJson;
import com.dslplatform.json.JsonReader;
import com.dslplatform.json.MapConverter;

import org.slf4j.Logger;
import org.stagemonitor.configuration.source.AbstractConfigurationSource;

import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.util.Collections;
import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.TimeUnit;

import co.elastic.apm.android.common.internal.logging.Elog;
import co.elastic.apm.android.sdk.internal.configuration.Configurations;
import co.elastic.apm.android.sdk.internal.configuration.impl.ConnectivityConfiguration;
import co.elastic.apm.android.sdk.internal.features.centralconfig.fetcher.CentralConfigurationFetcher;
import co.elastic.apm.android.sdk.internal.features.centralconfig.fetcher.ConfigurationFileProvider;
import co.elastic.apm.android.sdk.internal.features.centralconfig.fetcher.FetchResult;
import co.elastic.apm.android.sdk.internal.services.Service;
import co.elastic.apm.android.sdk.internal.services.ServiceManager;
import co.elastic.apm.android.sdk.internal.services.preferences.PreferencesService;
import co.elastic.apm.android.sdk.internal.time.SystemTimeProvider;

public final class CentralConfigurationManager extends AbstractConfigurationSource implements ConfigurationFileProvider {
    private static final String REFRESH_TIMEOUT_PREFERENCE_NAME = "central_configuration_refresh_timeout";
    private final Context context;
    private final DslJson<Object> dslJson = new DslJson<>(new DslJson.Settings<>());
    private final Map<String, String> configs = new HashMap<>();
    private final Logger logger = Elog.getLogger();
    private final byte[] buffer = new byte[4096];
    private final PreferencesService preferences;
    private final SystemTimeProvider systemTimeProvider;
    private File configFile;

    public CentralConfigurationManager(Context context) {
        this(context, SystemTimeProvider.get());
    }

    @VisibleForTesting
    public CentralConfigurationManager(Context context, SystemTimeProvider systemTimeProvider) {
        this.context = context;
        this.systemTimeProvider = systemTimeProvider;
        preferences = ServiceManager.get().getService(Service.Names.PREFERENCES);
    }

    public Integer sync() throws IOException {
        if (getRefreshTimeoutMillis() > systemTimeProvider.getCurrentTimeMillis()) {
            logger.debug("Ignoring central config sync request");
            return null;
        }
        try {
            CentralConfigurationFetcher fetcher = new CentralConfigurationFetcher(Configurations.get(ConnectivityConfiguration.class),
                    this, preferences);
            FetchResult fetchResult = fetcher.fetch();
            if (fetchResult.configurationHasChanged) {
                notifyListeners();
            }
            Integer maxAgeInSeconds = fetchResult.maxAgeInSeconds;
            if (maxAgeInSeconds != null) {
                storeRefreshTimeoutTime(maxAgeInSeconds);
            }
            return maxAgeInSeconds;
        } catch (Throwable t) {
            logger.error("An error occurred while fetching the central configuration", t);
            throw t;
        }
    }

    private void notifyListeners() throws IOException {
        configs.putAll(readConfigs(getConfigurationFile()));
        logger.info("Notifying central config change");
        logger.debug("Central config params: {}", configs);
        Configurations.reload();
        configs.clear();
    }

    private Map<String, String> readConfigs(File configFile) throws IOException {
        try (InputStream is = new FileInputStream(configFile)) {
            JsonReader<Object> reader = dslJson.newReader(is, buffer);
            reader.startObject();
            return Collections.unmodifiableMap(MapConverter.deserialize(reader));
        }
    }

    private void storeRefreshTimeoutTime(int maxAgeInSeconds) {
        logger.debug("Storing central config max age seconds {}", maxAgeInSeconds);
        setRefreshTimeoutMillis(systemTimeProvider.getCurrentTimeMillis() + TimeUnit.SECONDS.toMillis(maxAgeInSeconds));
    }

    private long getRefreshTimeoutMillis() {
        return preferences.retrieveLong(REFRESH_TIMEOUT_PREFERENCE_NAME, 0);
    }

    private void setRefreshTimeoutMillis(long timeoutMillis) {
        preferences.store(REFRESH_TIMEOUT_PREFERENCE_NAME, timeoutMillis);
    }

    @Override
    public File getConfigurationFile() {
        if (configFile == null) {
            configFile = new File(context.getFilesDir(), "elastic_agent_configuration.json");
        }
        return configFile;
    }

    public void publishCachedConfig() {
        File configurationFile = getConfigurationFile();
        if (!configurationFile.exists()) {
            logger.debug("No cached central config found");
            return;
        }
        try {
            notifyListeners();
        } catch (Throwable t) {
            logger.error("Exception when publishing cached central config", t);
        }
    }

    @Override
    public String getValue(String key) {
        return configs.get(key);
    }

    @Override
    public String getName() {
        return "APM Server";
    }
}

File: android-sdk/src/main/java/co/elastic/apm/android/sdk/features/centralconfig/CentralConfigurationManager.kt
/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *	http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package co.elastic.apm.android.sdk.features.centralconfig

import androidx.annotation.WorkerThread
import co.elastic.apm.android.common.internal.logging.Elog
import co.elastic.apm.android.sdk.connectivity.ConnectivityConfigurationHolder
import co.elastic.apm.android.sdk.features.apmserver.ApmServerConnectivityManager
import co.elastic.apm.android.sdk.features.exportergate.ExporterGateManager
import co.elastic.apm.android.sdk.internal.services.ServiceManager
import co.elastic.apm.android.sdk.internal.time.SystemTimeProvider
import java.io.IOException
import java.util.concurrent.TimeUnit
import org.slf4j.Logger
import org.stagemonitor.configuration.ConfigurationRegistry

class CentralConfigurationManager private constructor(
    serviceManager: ServiceManager,
    private val configurationRegistry: ConfigurationRegistry,
    private val centralConfigurationSource: CentralConfigurationSource,
    private val connectivityHolder: ConnectivityHolder,
    private val gateManager: ExporterGateManager
) : CentralConfigurationSource.Listener {
    private val backgroundWorkService by lazy { serviceManager.getBackgroundWorkService() }
    private val logger: Logger = Elog.getLogger()

    fun getConnectivityConfiguration(): CentralConfigurationConnectivity {
        return connectivityHolder.get() as CentralConfigurationConnectivity
    }

    fun setConnectivityConfiguration(configuration: CentralConfigurationConnectivity) {
        connectivityHolder.unlinkFromExportersConfig()
        connectivityHolder.set(configuration)
    }

    internal fun initialize() {
        backgroundWorkService.submit {
            try {
                centralConfigurationSource.initialize()
                openLatch()
                doPoll()
            } catch (t: Throwable) {
                logger.error("CentralConfiguration initialization error", t)
                scheduleDefault()
            } finally {
                openLatch()
            }
        }
    }

    internal fun getCentralConfiguration(): CentralConfiguration {
        return configurationRegistry.getConfig(CentralConfiguration::class.java)
    }

    private fun openLatch() {
        gateManager.openLatches(CentralConfigurationManager::class.java)
    }

    private fun scheduleDefault() {
        scheduleInSeconds(DEFAULT_POLLING_INTERVAL_IN_SECONDS)
    }

    private fun scheduleInSeconds(seconds: Int) {
        backgroundWorkService.scheduleOnce(
            TimeUnit.SECONDS.toMillis(seconds.toLong()),
            ConfigurationPoll()
        )
    }

    @WorkerThread
    @Throws(IOException::class)
    private fun doPoll() {
        val delayForNextPollInSeconds = centralConfigurationSource.sync()
        if (delayForNextPollInSeconds != null) {
            logger.info("Central config returned max age is null")
            scheduleInSeconds(delayForNextPollInSeconds)
        } else {
            scheduleDefault()
        }
    }

    companion object {
        private const val DEFAULT_POLLING_INTERVAL_IN_SECONDS = 60

        internal fun create(
            serviceManager: ServiceManager,
            systemTimeProvider: SystemTimeProvider,
            gateManager: ExporterGateManager,
            serviceName: String,
            serviceDeployment: String?,
            connectivityHolder: ApmServerConnectivityManager.ConnectivityHolder
        ): CentralConfigurationManager {
            val centralConfigurationConnectivityHolder = ConnectivityHolder.fromApmServerConfig(
                serviceName, serviceDeployment, connectivityHolder
            )
            val centralConfigurationSource = CentralConfigurationSource(
                serviceManager,
                centralConfigurationConnectivityHolder,
                systemTimeProvider
            )
            val registry = ConfigurationRegistry.builder()
                .addConfigSource(centralConfigurationSource)
                .addOptionProvider(CentralConfiguration())
            val latchName = "Central configuration"
            gateManager.createSpanGateLatch(CentralConfigurationManager::class.java, latchName)
            gateManager.createLogRecordLatch(CentralConfigurationManager::class.java, latchName)
            gateManager.createMetricGateLatch(CentralConfigurationManager::class.java, latchName)
            val centralConfigurationManager = CentralConfigurationManager(
                serviceManager,
                registry.build(),
                centralConfigurationSource,
                centralConfigurationConnectivityHolder,
                gateManager
            )
            centralConfigurationSource.listener = centralConfigurationManager
            return centralConfigurationManager
        }
    }

    private inner class ConfigurationPoll : Runnable {
        override fun run() {
            try {
                doPoll()
            } catch (t: Throwable) {
                logger.error("Central config poll error", t)
                scheduleDefault()
            }
        }
    }

    internal class ConnectivityHolder(
        connectivity: CentralConfigurationConnectivity,
        private val apmServerConnectivityHolder: ApmServerConnectivityManager.ConnectivityHolder,
        private val serviceName: String,
        private val serviceDeployment: String?
    ) : ConnectivityConfigurationHolder(connectivity), ConnectivityConfigurationHolder.Listener {

        companion object {
            fun fromApmServerConfig(
                serviceName: String,
                serviceDeployment: String?,
                apmServerConfigurationManager: ApmServerConnectivityManager.ConnectivityHolder
            ): ConnectivityHolder {
                val instance = ConnectivityHolder(
                    CentralConfigurationConnectivity.fromApmServerConfig(
                        serviceName,
                        serviceDeployment,
                        apmServerConfigurationManager.getConnectivityConfiguration()
                    ), apmServerConfigurationManager, serviceName, serviceDeployment
                )
                apmServerConfigurationManager.addListener(instance)
                return instance
            }
        }

        internal fun unlinkFromExportersConfig() {
            apmServerConnectivityHolder.removeListener(this)
        }

        override fun onConnectivityConfigurationChange() {
            // On apm server exporters config change.
            set(
                CentralConfigurationConnectivity.fromApmServerConfig(
                    serviceName,
                    serviceDeployment,
                    apmServerConnectivityHolder.getConnectivityConfiguration()
                )
            )
        }
    }

    override fun onConfigChange() {
        configurationRegistry.reloadDynamicConfigurationOptions()
    }
}